validator/validator.go

```diff
@@ -2,12 +2,53 @@ package validator
 
 import (
 	"errors"
+
+	"github.com/GridPlus/phonon-client/chain"
 	"github.com/GridPlus/phonon-client/model"
 )
 
-var ErrMissingPubKey = errors.New("phonon missing public key")
+var ErrUnsupportedCurrency = errors.New("unsupported currency type")
+var ErrEmptyProposal = errors.New("there were no phonon(s) to validate")
+
+type AssetValidationResult struct {
+	P     *model.Phonon
+	Valid bool
+	Err   error
+}
+
+type AssetValidator interface {
+	Validate(proposal []*model.Phonon) ([]*AssetValidationResult, error)
+}
+
+type Validator struct {
+	ethValidator   *chain.EthChainService
+	assetValidator AssetValidator
+}
+
+func NewValidator(assetValidator AssetValidator) *Validator {
+	return &Validator{assetValidator: assetValidator}
+}
+
+func (v *Validator) Validate(proposal []*model.Phonon) ([]*AssetValidationResult, error) {
+	result := []*AssetValidationResult{}
+
+	if len(proposal) == 0 {
+		return nil, ErrEmptyProposal
+	}
+
+	for _, p := range proposal {
+		switch p.CurrencyType {
+		case 2:
+			result, err := v.ethValidator.Validate(p)
```

Comment:
Does this loop work? I think initializing result inside the for loop here will erase the contents of result from the last iteration. Not what we want. Think this is a good case to use named return parameters. 

```suggestion
//In the method sig
func (v *Validator) Validate(proposal []*model.Phonon) (results []*AssetValidationResult, err error) {
...
case 2: 
result, err := v.ethValidator.Validate(p)
...
results = append(results, result)
```